README.rst

```diff
@@ -0,0 +1,297 @@
+
+ - If the `namespace` is empty:
+
+   - Append the percent-encoded name to the `puurl`
+
+ - If the `version` is not empty:
+
+   - Append '@' to the `puurl`
+   - Append the percent-encoded version to the `puurl`
+
+ - If the `qualifiers` are not empty:
+
+   - Append '?' to the `puurl`
+   - For each key/value pair:
+   
+     - create a string by joining the lowercase `key`, the equal '=' 
+       sign and the percent-encoded `value`
+
+   - sort this list of strings lexicographically
+   - join this list of strings with a '&'
+   - Append this to the `puurl`
+
+ - If the `path` is not empty:
+ 
+   - Append '#' to the `puurl`
+   - Split the `path` on '/'
+   - Discard empty, '.' and '..' segments
+   - Percent-encode each segment
+   - Join the segments with '/'
+   - Append this to the `puurl`
+
+
+How to parse a `puurl` string in its parts
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Based on the conventions defined in this document, parsing works from
+right to left.
+
+- Split the `puurl` string once from right on '#'
+ 
+  - The left side is the `remainder`
+  - Strip the right side from leading and trailing '/'
+  - Split this on '/'
+  - Discard any empty string segment from that split
+  - Discard any '.' or  '..' segment from that split
+  - Percent-decode each segment and join them back in a '/'
+    slash-separated string. This is the `path`
+
+- Split the `remainder` once from right on '?'
+ 
+  - The right side is the `qualifiers`
+  - The left side is the `remainder`
+  - Split the `qualifiers` on '&'. Each part is a `key=value` pair
+  - For each pair, split the `key=value` once from left on '=':
+
+    - The `key` is the lowercase left side
+    - The `value` is the percent-decoded right side
+
+- Split the `remainder` once from left on ':'
+ 
+  - The left side lowercased is the `type`
+  - The right side is the `remainder`
+
+- Strip the `remainder` from leading and trailing '/'
+ 
+  - Split this once from right on '/'
+  - The right side is the `name` after percent-decoding
+  - The left side is the `remainder`
+
+- Split the `remainder` on '/'
+ 
+  - Discard any empty segment from that split
+  - Percent-decode each segment and join them back in a '/'
+    slash-separated string
+  - This is the `namespace`
+
+
+Relations with the whatwg URLs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These are the relationships between the parts of a `puurl` and a
+URL as defined at https://url.spec.whatwg.org/#url-representation :
+
+- `puurl` `type`: this is the URL `scheme`.
+- `puurl` `namespace`, `name` and `version` parts: these collectively map to the URL `path`. 
+- `puurl` `qualifiers`: this maps to the URL `query`
+- `puurl` `path`: this is the URL `fragment`
+
+Note: In a `puurl` there is no special mapping of parts to the URL `username`,
+`password`, `host` and `port` parts.
+
+
+Known implementations
+~~~~~~~~~~~~~~~~~~~~~
+
+- in Python:
+- in Ruby:
+- in Go:
+- in JavaScript:
+- in Perl:
+- for the JVM:
```

Comment:
Please consider including .NET as well.

For implementation, these places are potential candidate for discussion:

https://github.com/nuget/home
https://github.com/dotnet/home
https://github.com/dotnet/standard

Reply:
Very good point! I am adding it in the updated version that I am pushing shortly

Reply:
@kasper3 this has been addressed in the latest push.